[BUG] SegmentReplicationIT.testScrollCreatedOnReplica is flaky

I caught this by running all SegRep ITs on repeat - its possible the scroll/pit requests aren't immediately clearing unreferenced files, this can be fixed by wrapping in an assertBusy.

```java
REPRODUCE WITH: ./gradlew 'null' --tests "org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testScrollCreatedOnReplica" -Dtests.seed=B6DEF7E40AB8FEBD -Dtests.locale=pl -Dtests.timezone=America/Louisville -Druntime.java=19

java.lang.AssertionError: Files should be cleaned up post scroll clear request

	at __randomizedtesting.SeedInfo.seed([B6DEF7E40AB8FEBD:AB0D24631C17B1C8]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at org.opensearch.indices.replication.SegmentReplicationIT.testScrollCreatedOnReplica(SegmentReplicationIT.java:1076)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20
```

Reopening it, this issue is not fixed:

```
java.lang.AssertionError: Files should be cleaned up post scroll clear request
	at __randomizedtesting.SeedInfo.seed([A16D8358F150F61C:BCBE50DFE7FFB969]:0)
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
	at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1089)
	at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1062)
	at org.opensearch.indices.replication.SegmentReplicationIT.testScrollCreatedOnReplica(SegmentReplicationIT.java:1073)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
	Suppressed: java.lang.AssertionError: Files should be cleaned up post scroll clear request
		at org.junit.Assert.fail(Assert.java:89)
		at org.junit.Assert.assertTrue(Assert.java:42)
		at org.junit.Assert.assertFalse(Assert.java:65)
		at org.opensearch.indices.replication.SegmentReplicationIT.lambda$testScrollCreatedOnReplica$11(SegmentReplicationIT.java:1074)
		at org.opensearch.test.OpenSearchTestCase.assertBusy(OpenSearchTestCase.java:1077)
		... 40 more
```

https://build.ci.opensearch.org/job/gradle-check/30636/testReport/junit/org.opensearch.remotestore/SegmentReplicationUsingRemoteStoreIT/testScrollCreatedOnReplica/

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationBaseIT.java
@@ -8,6 +8,7 @@
 
 package org.opensearch.indices.replication;
 
+import org.apache.lucene.index.SegmentInfos;
 import org.opensearch.action.search.SearchResponse;
 import org.opensearch.cluster.ClusterState;
 import org.opensearch.cluster.metadata.IndexMetadata;
@@ -16,6 +17,8 @@ import org.opensearch.cluster.routing.IndexRoutingTable;
 import org.opensearch.cluster.routing.IndexShardRoutingTable;
 import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.common.Nullable;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.common.concurrent.GatedCloseable;
 import org.opensearch.common.lease.Releasable;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.core.index.Index;
@@ -28,12 +31,14 @@ import org.opensearch.index.shard.IndexShard;
 import org.opensearch.index.store.Store;
 import org.opensearch.index.store.StoreFileMetadata;
 import org.opensearch.indices.IndicesService;
+import org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint;
 import org.opensearch.indices.replication.common.ReplicationType;
 import org.opensearch.plugins.Plugin;
 import org.opensearch.test.OpenSearchIntegTestCase;
 import org.opensearch.test.transport.MockTransportService;
 import org.opensearch.transport.TransportService;
 
+import java.io.IOException;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
@@ -241,5 +246,15 @@ public class SegmentReplicationBaseIT extends OpenSearchIntegTestCase {
                 assertEquals(0, shardStat.getCheckpointsBehindCount());
             }
         }, 30, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Returns the latest SIS for a shard but does not incref the segments.
+     */
+    protected SegmentInfos getLatestSegmentInfos(IndexShard shard) throws IOException {
+        final Tuple<GatedCloseable<SegmentInfos>, ReplicationCheckpoint> tuple = shard.getLatestSegmentInfosAndCheckpoint();
+        try (final GatedCloseable<SegmentInfos> closeable = tuple.v1()) {
+            return closeable.get();
+        }
     }
 }

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -1079,8 +1079,16 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         // there should be no active readers, snapshots, or on-disk commits containing the snapshotted files, check that they have been
         // deleted.
         Set<String> latestCommitSegments = new HashSet<>(replicaShard.store().readLastCommittedSegmentsInfo().files(false));
-        assertEquals("Snapshotted files are no longer part of the latest commit", Collections.emptySet(), Sets.intersection(latestCommitSegments, snapshottedSegments));
-        assertEquals("All snapshotted files should be deleted", Collections.emptySet(), Sets.intersection(filesAfterClearScroll, snapshottedSegments));
+        assertEquals(
+            "Snapshotted files are no longer part of the latest commit",
+            Collections.emptySet(),
+            Sets.intersection(latestCommitSegments, snapshottedSegments)
+        );
+        assertEquals(
+            "All snapshotted files should be deleted",
+            Collections.emptySet(),
+            Sets.intersection(filesAfterClearScroll, snapshottedSegments)
+        );
     }
 
     /**

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -989,7 +989,13 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
     public void testScrollCreatedOnReplica() throws Exception {
         // create the cluster with one primary node containing primary shard and replica node containing replica shard
         final String primary = internalCluster().startDataOnlyNode();
-        createIndex(INDEX_NAME);
+        prepareCreate(
+            INDEX_NAME,
+            Settings.builder()
+                .put(indexSettings())
+                // we want to control refreshes
+                .put("index.refresh_interval", -1)
+        ).get();
         ensureYellowAndNoInitializingShards(INDEX_NAME);
         final String replica = internalCluster().startDataOnlyNode();
         ensureGreen(INDEX_NAME);
@@ -1032,9 +1038,9 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
                 .setId(String.valueOf(i))
                 .setSource(jsonBuilder().startObject().field("field", i).endObject())
                 .get();
+            refresh(INDEX_NAME);
         }
         // create new on-disk segments and copy them out.
-        refresh(INDEX_NAME);
         assertBusy(() -> {
             assertEquals(
                 getIndexShard(primary, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion(),

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -57,6 +57,7 @@ import org.opensearch.common.lease.Releasable;
 import org.opensearch.common.lucene.index.OpenSearchDirectoryReader;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.set.Sets;
 import org.opensearch.core.common.io.stream.NamedWriteableRegistry;
 import org.opensearch.core.index.shard.ShardId;
 import org.opensearch.core.xcontent.XContentBuilder;
@@ -89,6 +90,8 @@ import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -1029,7 +1032,7 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
 
         final IndexShard replicaShard = getIndexShard(replica, INDEX_NAME);
         SegmentInfos latestSegmentInfos = getLatestSegmentInfos(replicaShard);
-        Collection<String> snapshottedSegments = latestSegmentInfos.files(false);
+        final Set<String> snapshottedSegments = new HashSet<>(latestSegmentInfos.files(false));
         logger.info("Segments {}", snapshottedSegments);
 
         // index more docs and force merge down to 1 segment
@@ -1072,16 +1075,12 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         assertEquals(1, scrollHits);
 
         client(replica).prepareClearScroll().addScrollId(searchResponse.getScrollId()).get();
-        List<String> filesAfterClearScroll = List.of(replicaShard.store().directory().listAll());
+        final Set<String> filesAfterClearScroll = Arrays.stream(replicaShard.store().directory().listAll()).collect(Collectors.toSet());
         // there should be no active readers, snapshots, or on-disk commits containing the snapshotted files, check that they have been
         // deleted.
-        for (String file : snapshottedSegments) {
-            assertFalse(
-                "Latest on-disk commit does not contain snapshotted segments",
-                replicaShard.store().readLastCommittedSegmentsInfo().files(false).contains(file)
-            );
-            assertFalse("File " + file + " should no longer be on disk", filesAfterClearScroll.contains(file));
-        }
+        Set<String> latestCommitSegments = new HashSet<>(replicaShard.store().readLastCommittedSegmentsInfo().files(false));
+        assertEquals("Snapshotted files are no longer part of the latest commit", Collections.emptySet(), Sets.intersection(latestCommitSegments, snapshottedSegments));
+        assertEquals("All snapshotted files should be deleted", Collections.emptySet(), Sets.intersection(filesAfterClearScroll, snapshottedSegments));
     }
 
     /**

--- a/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
+++ b/server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java
@@ -53,8 +53,6 @@ import org.opensearch.cluster.routing.ShardRouting;
 import org.opensearch.cluster.routing.ShardRoutingState;
 import org.opensearch.cluster.routing.allocation.command.CancelAllocationCommand;
 import org.opensearch.common.action.ActionFuture;
-import org.opensearch.common.collect.Tuple;
-import org.opensearch.common.concurrent.GatedCloseable;
 import org.opensearch.common.lease.Releasable;
 import org.opensearch.common.lucene.index.OpenSearchDirectoryReader;
 import org.opensearch.common.settings.Settings;
@@ -73,7 +71,6 @@ import org.opensearch.index.engine.EngineConfig;
 import org.opensearch.index.engine.NRTReplicationReaderManager;
 import org.opensearch.index.shard.IndexShard;
 import org.opensearch.indices.recovery.FileChunkRequest;
-import org.opensearch.indices.replication.checkpoint.ReplicationCheckpoint;
 import org.opensearch.indices.replication.common.ReplicationType;
 import org.opensearch.node.NodeClosedException;
 import org.opensearch.search.SearchService;
@@ -997,27 +994,20 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
         final String replica = internalCluster().startDataOnlyNode();
         ensureGreen(INDEX_NAME);
 
-        // index 10 docs
-        for (int i = 0; i < 10; i++) {
-            client().prepareIndex(INDEX_NAME)
-                .setId(String.valueOf(i))
-                .setSource(jsonBuilder().startObject().field("field", i).endObject())
-                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
-                .get();
-            refresh(INDEX_NAME);
-        }
+        client().prepareIndex(INDEX_NAME)
+            .setId(String.valueOf(0))
+            .setSource(jsonBuilder().startObject().field("field", 0).endObject())
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .get();
+        refresh(INDEX_NAME);
+
         assertBusy(
             () -> assertEquals(
                 getIndexShard(primary, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion(),
                 getIndexShard(replica, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion()
             )
         );
-        final IndexShard replicaShard = getIndexShard(replica, INDEX_NAME);
-        final Tuple<GatedCloseable<SegmentInfos>, ReplicationCheckpoint> tuple = replicaShard.getLatestSegmentInfosAndCheckpoint();
-        final Collection<String> snapshottedSegments;
-        try (final GatedCloseable<SegmentInfos> closeable = tuple.v1()) {
-            snapshottedSegments = closeable.get().files(false);
-        }
+
         // opens a scrolled query before a flush is called.
         // this is for testing scroll segment consistency between refresh and flush
         SearchResponse searchResponse = client(replica).prepareSearch()
@@ -1031,17 +1021,20 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
             .setScroll(TimeValue.timeValueDays(1))
             .get();
 
-        // force call flush
-        flush(INDEX_NAME);
+        final IndexShard replicaShard = getIndexShard(replica, INDEX_NAME);
+        SegmentInfos latestSegmentInfos = getLatestSegmentInfos(replicaShard);
+        Collection<String> snapshottedSegments = latestSegmentInfos.files(false);
+        logger.info("Segments {}", snapshottedSegments);
 
-        for (int i = 3; i < 5; i++) {
-            client().prepareDelete(INDEX_NAME, String.valueOf(i)).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
-            refresh(INDEX_NAME);
-            if (randomBoolean()) {
-                client().admin().indices().prepareForceMerge(INDEX_NAME).setMaxNumSegments(1).setFlush(true).get();
-                flush(INDEX_NAME);
-            }
+        // index more docs and force merge down to 1 segment
+        for (int i = 1; i < 5; i++) {
+            client().prepareIndex(INDEX_NAME)
+                .setId(String.valueOf(i))
+                .setSource(jsonBuilder().startObject().field("field", i).endObject())
+                .get();
         }
+        // create new on-disk segments and copy them out.
+        refresh(INDEX_NAME);
         assertBusy(() -> {
             assertEquals(
                 getIndexShard(primary, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion(),
@@ -1049,13 +1042,19 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
             );
         });
 
+        // force merge and flush.
         client().admin().indices().prepareForceMerge(INDEX_NAME).setMaxNumSegments(1).setFlush(true).get();
+        // wait for replication to complete
         assertBusy(() -> {
             assertEquals(
                 getIndexShard(primary, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion(),
                 getIndexShard(replica, INDEX_NAME).getLatestReplicationCheckpoint().getSegmentInfosVersion()
             );
         });
+        logger.info("Local segments after force merge and commit {}", getLatestSegmentInfos(replicaShard).files(false));
+        List<String> filesBeforeClearScroll = List.of(replicaShard.store().directory().listAll());
+        assertTrue("Files should be preserved", filesBeforeClearScroll.containsAll(snapshottedSegments));
+
         // Test stats
         logger.info("--> Collect all scroll query hits");
         long scrollHits = 0;
@@ -1064,20 +1063,19 @@ public class SegmentReplicationIT extends SegmentReplicationBaseIT {
             searchResponse = client(replica).prepareSearchScroll(searchResponse.getScrollId()).setScroll(TimeValue.timeValueDays(1)).get();
             assertAllSuccessful(searchResponse);
         } while (searchResponse.getHits().getHits().length > 0);
-
-        List<String> currentFiles = List.of(replicaShard.store().directory().listAll());
-        assertTrue("Files should be preserved", currentFiles.containsAll(snapshottedSegments));
+        assertEquals(1, scrollHits);
 
         client(replica).prepareClearScroll().addScrollId(searchResponse.getScrollId()).get();
-
-        assertBusy(
-            () -> assertFalse(
-                "Files should be cleaned up post scroll clear request",
-                List.of(replicaShard.store().directory().listAll()).containsAll(snapshottedSegments)
-            )
-        );
-        assertEquals(10, scrollHits);
-
+        List<String> filesAfterClearScroll = List.of(replicaShard.store().directory().listAll());
+        // there should be no active readers, snapshots, or on-disk commits containing the snapshotted files, check that they have been
+        // deleted.
+        for (String file : snapshottedSegments) {
+            assertFalse(
+                "Latest on-disk commit does not contain snapshotted segments",
+                replicaShard.store().readLastCommittedSegmentsInfo().files(false).contains(file)
+            );
+            assertFalse("File " + file + " should no longer be on disk", filesAfterClearScroll.contains(file));
+        }
     }
 
     /**

--- a/server/src/main/java/org/opensearch/index/engine/ReplicaFileTracker.java
+++ b/server/src/main/java/org/opensearch/index/engine/ReplicaFileTracker.java
@@ -17,7 +17,7 @@ import java.util.HashSet;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
-import java.util.function.BiConsumer;
+import java.util.function.Consumer;
 
 /**
  * This class is heavily influenced by Lucene's ReplicaFileDeleter class used to keep track of
@@ -31,10 +31,10 @@ final class ReplicaFileTracker {
 
     public static final Logger logger = LogManager.getLogger(ReplicaFileTracker.class);
     private final Map<String, Integer> refCounts = new HashMap<>();
-    private final BiConsumer<String, String> fileDeleter;
+    private final Consumer<String> fileDeleter;
     private final Set<String> EXCLUDE_FILES = Set.of("write.lock");
 
-    public ReplicaFileTracker(BiConsumer<String, String> fileDeleter) {
+    public ReplicaFileTracker(Consumer<String> fileDeleter) {
         this.fileDeleter = fileDeleter;
     }
 
@@ -82,7 +82,7 @@ final class ReplicaFileTracker {
 
     private synchronized void delete(String fileName) {
         assert canDelete(fileName);
-        fileDeleter.accept("delete unreferenced", fileName);
+        fileDeleter.accept(fileName);
     }
 
     private synchronized boolean canDelete(String fileName) {

